### DISCUSSION TITLE
Mason offline (or portable Mason?)

### DISCUSSION THREAD

**Discussion by smprather** [mason-org/mason.nvim#1526] [Q&A]
I did some searching. Found Gongfarmer talking about a corp proxy (and fixes that went way over my head). The firewall I'm working behind is a brick wall (zero internet access). I am trying to come up with a LazyVim/Mason config on an internet visible machine (WSL/Ubuntu) and then transfer over to the brick-walled environment (Redhat8) and have it Just Work™️. I can do updates on my internet machine. And I can transfer files over the brick-wall, but that's it.

I just tried rsync'ing everything over (including ~/.local/share/nvim), and as expected, ran into something. When ruff-lsp is installed, a python venv is created (based on /usr/bin/python3). Those symlinks are no longer valid on the Redhat system. The specifics are irrelevant. I knew something would come up.

The way I get around it with LazyVim.
1. I create a local git clone of every plugin I am referencing (they are legion, I have a script to update them all)
2. In LazyVim config, git = {url_format = "file:///home/epramyl/nvimide/plugin_src/%s"}
3. Rsync all of the plugin clones over the wall - effectively pinning the versions. That's fine. I can update on the WSL side when I want to. Then script-rsync it all over the wall.

Can I do the same in Mason? I need Mason to reference my local file:// urls that I will pre-establish with git clones (git clone --mirror). Then the install steps will be followed on the other side of the wall upon first invocation.

Thanks!
--Myles

[I will just vent a little: It has been very frustrating with the entire nvim ecosystem assuming that every invocation of nvim will have internet visibility. I'm pretty sure offline is a very common thing.]

**Comment by smprather** [comment]
I just looked at some Mason registry entries. It's becoming obvious to me that Mason is, by its very nature, _extremely_ "online". I understand that, and how difficult it would be to change.

So, how about making my Mason install dir portable? I need to be able to create it in one context (internet visible Ubuntu), and then airlift it into a different context (no internet Redhat). For example, I see that the sym-links created in mason/bin are absolute, but they could be relative (or will this break something, like Windows compatibility?). But ok, I could script-hack that. What's really getting me is the python LSP/etc are installing themselves probably with pipx which creates a venv with links of absolute paths to the python on my "build side". Well, I guess I could script hack those too (I am including a full Python env in my nvim build dir). It's looking like I need to create a sym-link relative'izer to run on the full dir tree.

Of course, there are binaries. I may be lucky here. I think binaries built in my WSL/Ubuntu are compatible with Redhat 8. I just copied the ruff binary over to the Redhat system and it runs ok (I'll deal with missing .so if/as needed).

I feel like I'm wandering down a doomed path. Ie, 10 hours of work and then I find that one little thing I didn't think of and can't solve.

Anyone have any comments/ideas?

**Comment by smprather** [comment]
Looks like I'm on my own. Fair 'nuf. I'll start by changing all sym-links to relative and see where I stand after that.

**Reply by qlibp** [reply]
you are not alone

**Comment by qlibp** [comment]
Interesting idea, and that's exactly what I'm trying to achieve too.

Seems that Mason2.0 provides a relative symlink, which works fine for clangd binary, but not python venv. ("Works fine" means that I could copy the whole mason directory to another machine and the lsp still work)

python venv is not quite portable, as all executable script has a absolute path shebang, so, when copying the whole venv to another machine, we can't use it directly & correctly, even we got the same python version.

For python venv, in order to make it portable, we need to use `pip download` & `pip install`([check this](https://pip.pypa.io/en/stable/user_guide/#installing-from-local-packages)). Currently mason exposed only the pip install api configuration, and I can't find a way to get a portable venv... Sad.

**Reply by qlibp** [reply]
For now, the script trick I could come up is:

**Computer with network**: `Mason Install python-lsp-server`->source python-lsp-server venv->`pip freeze -r requirement.txt`->`pip download -r requirement.txt` -> remove the whole venv except the newly downloaded archive

**Computer without network**: copy the whole nvim plugins directory->create python-lsp-server venv->`pip install --no-index --find-files /path/to/downloaded/archive -r requirement.txt`

This method should works fine, for all pyhton venv

**Reply by smprather** [reply]
When python creates a venv, the bin/ dir has absolute path links to the python install that was used as the base for the venv. This is not portable.  Here's my script to change the links to the python venv I'm packaging with my (hopefully) self-contained and portable nvimide distribution (my marketing name is not nearly as clever as all the others).

```python
#!./python/squashfs-root/usr/bin/python3
from pathlib import Path
import sys
import os

orig_cwd = Path.cwd()
cwd_parts = orig_cwd.resolve().parts

for file in orig_cwd.glob("**/*"):
    if not file.is_symlink():
        continue

    target_orig = file.readlink()
    target = file.readlink()
    if not target.is_absolute():
        continue

    print("Found symlink with absolute path target:")
    print(f"  Symlink: {file}")
    print(f"  Tartget: {target_orig}")

    for dir_part in cwd_parts:
        if target.parts[0] == dir_part:
            target = Path(*(target.parts[1:]))
        else:
            sys.exit(f"Error, target not in nvimide. File: {file}, target: {target}")

    if target.exists():
        rel_path = Path("../" * (len(file.parts) - 4 - 1)) / target
        print(f"  Changing link to rel-path: {rel_path}")
        os.chdir(file.parent)
        if not rel_path.exists():
            sys.exit("Error")
        Path(file.name).unlink()
        Path(file.name).symlink_to(rel_path)
        os.chdir(orig_cwd)
```

**Reply by qlibp** [reply]
I find out that lua_ls also use absolute path… do you get it to work? I don’t quite familiar  with the installation procedure of lua_ls

**Comment by williamboman** [comment]
Hello! I'm not sure I follow how mason.nvim would function without internet access? The only time internet access is needed is when installing a package, once installed it's made available regardless of whether you have internet access or not.

Copying the entire Mason directory from another system is not really supported mainly because of absolute links (https://github.com/williamboman/mason.nvim/issues/1156), which will be changed to relative in v2.0. As you've mentioned there may also be executables created during installation that use absolute paths (Python venv executables for instance). If you do this you'll want to install the packages on an identical system with glibc compatibility.

Tbh I'd recommend perhaps not using Mason if you're on such a closed off system (which probably is closed off for good reason) and instead use alternatives.

**Reply by smprather** [reply]
Think of it this way: I want to use a "normal" internet enabled NeoVim space to build my IDE. Then I want to transfer it into closed space, very preferably as a single directory not scattered across ~/.local, ~/.cache, ~/.etc.etc.etc. I will update on the build side from time to time, then transfer to closed space again. I will be sharing this custom nvim with other users. If one of them needs new LSP support (or other utility type nicely managed by Mason), then I add it on the build side and mirror to closed space. I'm surprised my use model seems to be so niche. But I guess it is. And since it is, it is reasonable to except some assembly required.

**Reply by smprather** [reply]
Also, after looking at a few kitchen sink alternatives, I've decided to use LazyVim as my starting point. LV is pretty "all in" on Mason. I would hate to have to try to cut it out.

**Reply by qlibp** [reply]
Actually, as I mentioned previously, pip allow us to download and then install, with this model, we can post install when migrate to other machine, and run install on that machine. Maybe expose this api for user to decide whether to preserve the download objects after installation?

**Comment by qlibp** [comment]
@williamboman Need some help on changing the generated absolute path of `lua-lsp-server` script.

Is the `lua-lsp-server` script generated by mason?
If so, which function is used to generate the script?

Is it possible to hack around the script-generation procedure, so that i can change the absolute path?
If not, is it only possible to directly hack the `lua-lsp-server` script after it was generated?

**Comment by smprather** [comment]
@williamboman I have another offline issue. Mason is hard-stopping my :checkhealth while offline. It won't even proceed after a timeout. Is there a way to a) have it respect a timeout, and b) set the timeout value myself (I will set it very small, like 1s)?

**Reply by xstraven** [reply]
Also embarking on this journey with one online and one offline machine. have there been any further developments?

**Reply by smprather** [reply]
I'm sorry to say that NeoVim is just going more and more and more online only. And you can see by @williamboman lack of response, offline is not of interest to the broader community. I understand nobody is getting paid to do any of this, and I would do the same on weekends - build to 95% of use-cases. But we need to accept the truth of the matter and try to come up with a plan to handle it.

Personally, I have a flow where I clone the entire github repos of ALL my plugins and ALL of the TreeSitter source code to my online environment. The I rsync it all over to offline space. For both TSInstall and LazyVim, I use their existing configs to swap https:// to file://my/path.

But to support Mason, we would have to change the URLs that are found in the source code.

**Reply by smprather** [reply]
I did just find this, so maybe there's hope.
```
github = {
            ---@since 1.0.0
            -- The template URL to use when downloading assets from GitHub.
            -- The placeholders are the following (in order):
            -- 1. The repository (e.g. "rust-lang/rust-analyzer")
            -- 2. The release version (e.g. "v0.3.0")
            -- 3. The asset name (e.g. "rust-analyzer-v0.3.0-x86_64-unknown-linux-gnu.tar.gz")
            download_url_template = "https://github.com/%s/releases/download/%s/%s",
        },
```

**Reply by xstraven** [reply]
thanks for the reply!
